core/protocols/smithy-rpcv2-protocol/src/main/java/software/amazon/awssdk/protocols/rpcv2/internal/SdkRpcV2CborGenerator.java

```diff
@@ -0,0 +1,120 @@
+import software.amazon.awssdk.thirdparty.jackson.core.JsonFactory;
+import software.amazon.awssdk.thirdparty.jackson.dataformat.cbor.CBORGenerator;
+
+/**
+ * Thin wrapper around Jackson's JSON generator for CBOR.
+ */
+@SdkInternalApi
+public final class SdkRpcV2CborGenerator extends SdkJsonGenerator {
+
+    private static final int CBOR_TAG_TIMESTAMP = 1;
+    private static final int MILLI_SECOND_PRECISION = 3;
+
+    SdkRpcV2CborGenerator(JsonFactory factory, String contentType) {
+        super(factory, contentType);
+    }
+
+    /**
+     * Jackson doesn't have native support for timestamp. As per the RFC 7049 (https://tools.ietf.org/html/rfc7049#section-2.4.1)
+     * we will need to write a tag and write the epoch.
+     */
+    @Override
+    public StructuredJsonGenerator writeValue(Instant instant) {
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeTag(CBOR_TAG_TIMESTAMP);
+            BigDecimal dateValue = BigDecimal.valueOf(instant.toEpochMilli());
+            generator.writeNumber(dateValue.scaleByPowerOfTen(0 - MILLI_SECOND_PRECISION).doubleValue());
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    public StructuredJsonGenerator writeStartArray(int size) {
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeStartArray(null, size);
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    public StructuredJsonGenerator writeValue(double val) {
+        if (canConvertToLong(val)) {
+            return writeValue((long) val);
+        }
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeNumber(val);
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    public StructuredJsonGenerator writeValue(float val) {
+        if (canConvertToLong(val)) {
+            return writeValue((long) val);
+        }
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeNumber(val);
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    protected CBORGenerator getGenerator() {
+        return (CBORGenerator) super.getGenerator();
+    }
+
+    /**
+     * Checks if we can convert the floating point value to a long. If we can we then use the writeNumber(long), alongside with
+     * having enabled Feature.WRITE_MINIMAL_INTS will allow us to represent the floating point values minimally.
+     */
+    private static boolean canConvertToLong(double value) {
+        return value % 1 == 0
+               && value >= Long.MIN_VALUE
+               && value <= Long.MAX_VALUE;
+    }
```

Comment:
How did we decide to do these optimizations?

Reply:
There was not a formal decision. The idea is trying to follow this bit [of the spec](https://smithy.io/2.0/additional-specs/protocols/smithy-rpc-v2.html#numeric-type-serialization)

> Floating-point numeric types (float and double) MAY be serialized into non-floating-point numeric types (byte, short, integer, and long) if and only if the conversion would not cause a loss of precision. For example, a field modeled as a float with a value of 256 may be sent as an integer.

Let me know if you know a better way, or if you'd prefer not to implement this bit.

Reply:
It would be an interesting thing to benchmark. Do we know of services with a lot of doubles that are usually integers?